Welcome to the ConkatSeq Documentation Manual
=============================================

#### ConkatSeq is a python program for concatenating amplicon sequencing data back to its chromosomal location in the genome. 

Table of Contents
-----------------

- [Description](#about)
  - [Concept](#concept)
  - [Algorithm](#algorithm)
- [Input](#input)
- [Output](#output)
- [Installation](#installation)
  - [Hardware dependencies](#hardware)
  - [Software dependencies](#software)
  - [Install](#install)
  
- [Usage](#usage)
  1. [Generate clustering table](#table)
  2. [Polish clustering table](#polish)
  3. [Compute and Graph clustering table](#graph)

- [Example](#example)
  

## <a name="about"></a> Description

### <a name="concept"></a> Concept

### <a name="algorithm"></a> Algorithm

## <a name="input"></a> Input

Illumina amplicon sequencing data file(s) in FASTA format.

### NOTE: The amplicon sequencing data file(s) must be demultiplexed or split into its corresponding barcoded numbers of well plate. 

#### Ex: If the sequencing was performed on a 384 well plate, the demultiplexed data will consist of 384 individual FASTA files representing each barcoded sample well in the plate.

## <a name="output"></a> Output

Network graph of highly associated coccurence clusters in GraphML format. 

## <a name="input"></a> Installation

### <a name="hardware"></a> Hardware Dependencies

ConkatSeq works in the following operating systems:

  -MacOS
  -Unix
  

### <a name="hardware"></a> Software Dependencies

## <a name="input"></a> Usage

## <a name="input"></a> Example


https://rockefeller.app.box.com/s/rhrgw13ux6qdns0vax5i4cgyvgtfdyd7




